### PR TITLE
Implement a `removeClass()` option in addition to `addClass()` in src/View/StringTemplate.php

### PR DESCRIPTION
**Pull request by groovenectar**
Implement the option to remove default classes as well as adding them. Uses similar logic as `addClass()`, except that it will `array_diff()` instead of `array_merge()`

A example case where this comes in handy are custom Bootstrap helpers. Perhaps a button usually gets class `btn-primary` assigned by default, and it may be desirable to remove that in order to use a different or blank button class

### CODE REVIEW

**Comment by dereuromark** [comment]
This requires some test cases.

File: src/View/StringTemplate.php [hunk 1]
```diff
@@ -369,4 +369,46 @@ public function addClass($input, $newClass, $useIndex = 'class')
...
+     *
+     * @param array|string $input The array or string to remove the class from
+     * @param array|string $removeClass the new class or classes to remove
+     * @param string $useIndex if you are inputting an array with an element other than default of 'class'.
+     * @return array|string
+     */
+    public function removeClass($input, $removeClass, $useIndex = 'class')
+    {
+        // NOOP
+        if (empty($removeClass)) {
```
Comment (markstory) [inline src/View/StringTemplate.php]: What if the class is `'0'`?

Comment (ADmad) [reply src/View/StringTemplate.php]: CSS identifiers can't start with a digit, so I don't think it's a problem.

https://www.w3.org/TR/CSS21/syndata.html#characters

File: src/View/StringTemplate.php [hunk 2]
```diff
@@ -369,4 +369,46 @@ public function addClass($input, $newClass, $useIndex = 'class')
 
         return $input;
     }
+
+    /**
+     * Removes a class and returns a unique list either in array or space separated
+     *
+     * @param array|string $input The array or string to remove the class from
+     * @param array|string $removeClass the new class or classes to remove
```
Comment (dereuromark) [inline src/View/StringTemplate.php]: array could be string[] here, right?

File: src/View/StringTemplate.php [hunk 3]
```diff
@@ -369,4 +369,46 @@ public function addClass($input, $newClass, $useIndex = 'class')
...
+     * @param string $useIndex if you are inputting an array with an element other than default of 'class'.
+     * @return array|string
```
Comment (dereuromark) [inline src/View/StringTemplate.php]: Same here probably.

**Comment by groovenectar** [comment]
> This requires some test cases.

That makes sense! I overlooked this. Hopefully will be able to get those in there within a week

File: src/View/StringTemplate.php (hunk 2 above)
Comment (groovenectar) [reply src/View/StringTemplate.php]: Yep, I think if it's an array, it's supposed to only be string[].. The `addClass()` one seems pretty flexible and intuitive, and `removeClass()` is using the same methodology

**Comment by groovenectar** [comment]
This is my first PR for Cake that is an update to something testable

File: src/View/StringTemplate.php (hunk 2 above)
Comment (groovenectar) [reply src/View/StringTemplate.php]: Just verified that that it does work like `addClass()` in that it supports both string and string[], I do need to add official test cases though

File: src/View/StringTemplate.php (hunk 3 above)
Comment (groovenectar) [reply src/View/StringTemplate.php]: These would be the same with `addClass()`

When I get back to this, should I change those to:

`* @param array string[]|string`

or

`* @param string[]|string`

Or is there a particular format?

Comment (ADmad) [reply src/View/StringTemplate.php]: `@param string[]|string`

Comment (groovenectar) [reply src/View/StringTemplate.php]: This is updated, and I'll work on getting tests up and running ASAP

**Comment by dereuromark** [comment]
ping @groovenectar 
Are you able to provide some test cases for us here? So we can merge this?

**Comment by groovenectar** [comment]
Sorry for my delay on this @markstory. I just now looked for tests on `addClass()` to use as an example, and it seems there aren't any tests for that either unless I'm missing something...

**Comment by dereuromark** [comment]
It is possible, best to add some for your new method.

File: tests/TestCase/View/StringTemplateTest.php [hunk 4]
```diff
@@ -440,4 +443,98 @@ public function testAddClassMethodUseIndex()
...
+        $result = $this->template->removeClass('keep_class remove_class', 'remove_class');
+        $this->assertEquals($result, ['class' => ['keep_class']]);
+
+        $result = $this->template->removeClass(null, 'remove_class');
+        $this->assertEquals($result, ['class' => []]);
+
+        $result = $this->template->removeClass(false, 'remove_class');
+        $this->assertEquals($result, ['class' => []]);
+
+        $result = $this->template->removeClass(new \StdClass(), 'remove_class');
```
Comment (dereuromark) [inline tests/TestCase/View/StringTemplateTest.php]: stdClass()

**Comment by groovenectar** [comment]
@dereuromark I found the tests, and have added some for the `removeClass()` method. There may be updates to the `addClass()` method upstream by now. I adapted the tests from `addClass()`. The main one I'm questioning is the last `assertEquals` test in the file... Normally I don't change that `$useIndex` parameter in practice...

Also, the original pull request may not be very DRY... I was thinking about adding another method to separate out the repeated parts of the two methods...

File: tests/TestCase/View/StringTemplateTest.php (hunk 4 above)
Comment (groovenectar) [reply tests/TestCase/View/StringTemplateTest.php]: @dereuromark, I updated this for both `addClass()` and `removeClass()`

**Comment by othercorey** [comment]
@groovenectar Did you delete the fork?

**Comment by groovenectar** [comment]
> @groovenectar Did you delete the fork?

Sorry 😬 I had to make another fork, and deleted the first one to do it

I guess it makes sense that it would yank it from PRs after that

I think that my tests were not super polished or comprehensive anyway for this, and the code could have been more DRY

I'm sure that I can find the method again or re-implement it. Until then, if it's gone I guess it would make sense to just close it. Whoops 😑

**Comment by groovenectar** [comment]
This was to go into Cake3, and now I'm setup to contribute to Cake4 anyway, so if I can re-implement or find it I'll make the request for version 4 instead

**Comment by othercorey** [comment]
You can view the files changed on this PR to see your work.

**Comment by groovenectar** [comment]
> You can view the files changed on this PR to see your work.

Ah, nice! That will help, thanks. Would it make more sense if I redo it for the latest 4.x branch?

**Comment by othercorey** [comment]
Yes. You can add it to 3.next or 4.x.  The 4.x branch is best and then can backport if needed.

**Pull request closed without merging**